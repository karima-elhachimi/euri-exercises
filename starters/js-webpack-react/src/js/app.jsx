import React from 'react';
import { hot } from 'react-hot-loader/root';
import jsImage from '../public/images/js-logo.png';

export function App() {
  return (
    <h1>
      <img src={jsImage} alt="javascript" height="100" />
      <span>Hello from ES2015+</span>
    </h1>
  );
}

export default hot(App);
