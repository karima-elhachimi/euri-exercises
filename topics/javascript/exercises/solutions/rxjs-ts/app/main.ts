import { Observable } from 'rxjs';

const observer = {
  next(val: any) {
    console.log(val);
  },
  error(err: any) {
    console.log('ERROR', err);
  },
  completed() {
    console.log('|');
  },
};

const btn = document.querySelector('#btn');

// Create a stream of a single element with type string
Observable.of('This is a mesage').subscribe(observer);

// Create a subscription that fires once, if you hover over a text element
const stream$ = Observable.fromEvent(btn!, 'mousemove');
stream$.take(1).subscribe(observer);

// Count the number of elements in a array (with an observable stream)
const array$ = Observable.from([1, 2, 3, 4]).count().subscribe(observer);

// Count the number of click events on a button
const clicks$ = Observable.fromEvent(btn!, 'click')
  .mapTo(1)
  .scan<number>((acc, val) => {
    return (acc = acc + 1);
  })
  .subscribe(observer);

// Create a count down timer in seconds (10 - 0)
Observable.interval(500).map(x => 10 - x).take(10).subscribe(observer);

// Create a observable stream of an input box value change
const input = document.querySelector('#input');
Observable.fromEvent<KeyboardEvent>(input!, 'keyup')
  .map(event => (event.target as HTMLInputElement).value)
  .filter(val => val.length > 2)
  .debounceTime(500)
  .map(val => val.toUpperCase())
  .subscribe(observer);

// Create a subscription on a button click that auto unsubscribe after 5 seconds
const timeout$ = Observable.timer(1000);
Observable.fromEvent(btn!, 'click').takeUntil(timeout$).subscribe(observer);
/*
    ---c--c---c------c----c-----
    ----------------0|
    ---c--c---c-----|
*/

// One subscribe with two buttons
const btnStart = document.querySelector('#btnStart');
const btnStop = document.querySelector('#btnStop');
const btnStart$ = Observable.fromEvent(btnStart!, 'click');
const btnStop$ = Observable.fromEvent(btnStop!, 'click');
btnStart$.merge(btnStop$).subscribe(observer);

// Start & stop timer (log to console) with a start and stop button
btnStart$
  .switchMap(event => Observable.interval(500).takeUntil(btnStop$))
  .subscribe(observer);
