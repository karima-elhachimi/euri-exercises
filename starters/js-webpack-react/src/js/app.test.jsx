import React from 'react';
import { render as renderRtl } from 'react-testing-library';

import { App } from './app';

describe('App', () => {
  function render() {
    return renderRtl(<App />);
  }

  it('it renders as expected', () => {
    const { container } = render();

    expect(container).toMatchSnapshot();
  });

  describe('preferred way of writing test which enables tdd', () => {
    test('it renders the js logo', () => {
      const { getByAltText } = render();

      const logo = getByAltText(/js logo/i);

      expect(logo).toHaveAttribute('src', 'public/images/js-logo.png');
    });

    test('it renders the welcome message', () => {
      const { getByTestId } = render();

      const welcomeMessage = getByTestId('welcome-message');

      expect(welcomeMessage).toHaveTextContent('Hello from ES2015+');
    });
  });
});
