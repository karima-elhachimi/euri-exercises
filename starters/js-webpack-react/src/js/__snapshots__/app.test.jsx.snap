// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App it renders as expected 1`] = `
<div>
  <h1>
    <img
      alt="JS Logo"
      height="100"
      src="public/images/js-logo.png"
    />
    <span
      data-testid="welcome-message"
    >
      Hello from ES2015+
    </span>
  </h1>
</div>
`;
